File: Hashing/python/intersection.py
from typing import Iterable, List, TypeVar

T = TypeVar("T")


def intersection(first: Iterable[T], second: Iterable[T]) -> List[T]:
    second_set = set(second)
    added = set()
    result: List[T] = []

    for item in first:
        if item in second_set and item not in added:
            added.add(item)
            result.append(item)

    return result

File: Hashing/java/intersection.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Intersection {
  public static <T> List<T> intersection(List<T> first, List<T> second) {
    Set<T> lookup = new HashSet<>(second);
    Set<T> added = new HashSet<>();
    List<T> result = new ArrayList<>();

    for (T item : first) {
      if (lookup.contains(item) && !added.contains(item)) {
        result.add(item);
        added.add(item);
      }
    }

    return result;
  }

  public static void run() {
    // intentionally empty
  }
}
